roxie/udplib/udpsha.hpp

```diff
@@ -126,65 +126,78 @@ class simple_queue
 {
     _et             *elements;
     unsigned int    element_count;
-    int             first;
-    int             last;
-    int             active_buffers;
+    unsigned        first = 0;
+    unsigned        last = 0;
+    unsigned        active_buffers = 0;
     CriticalSection c_region;
     Semaphore       data_avail;
     Semaphore       free_space;
     
 public: 
-    void push(const _et &element)
+    bool push(const _et &element, unsigned timeout = (unsigned)-1)
     {
-        free_space.wait();
-        c_region.enter();
-        int next = (last + 1) % element_count;
-        elements[last] = element;
-        last = next;
-        active_buffers++;
-        c_region.leave();
+        if (!free_space.wait(timeout))
+            return false;
+
+        {
+            CriticalBlock block(c_region);
+            doPush(element);
+        }
+
         data_avail.signal();
+        return true;
     }
-    
-    bool push(const _et &element,long timeout)
+
+    template <class FUNC>
+    bool pushOrModify(const _et &element, FUNC walkFunc, unsigned timeout = (unsigned)-1)
     {
-        if (free_space.wait(timeout) ) {
-            c_region.enter();
-            int next = (last + 1) % element_count;
-            elements[last] = element;
-            last = next;
-            active_buffers++;
-            c_region.leave();
-            data_avail.signal();
-            return true;
+        if (!free_space.wait(timeout))
+            return false;
+
+        {
+            CriticalBlock block(c_region);
+            if (doWalk(walkFunc))
+                return true;
+            doPush(element);
         }
-        return false;
+
+        data_avail.signal();
+        return true;
     }
+
     
     void pop (_et &element) 
     {
         data_avail.wait();
-        c_region.enter();
-        element = elements[first];
-        first = (first + 1) % element_count;
-        active_buffers--;
-        c_region.leave();
+
+        {
+            CriticalBlock block(c_region);
+            element = elements[first];
+            first = (first + 1) % element_count;
+            active_buffers--;
+        }
+
         free_space.signal();
     }
     
     unsigned in_queue() {
-        c_region.enter();
-        unsigned res = active_buffers;
-        c_region.leave();
-        return res;
+        CriticalBlock block(c_region);
```

Comment:
This (and the one below) feel like rather pointless critical blocks - would making active_buffers atomic be better?